### PR TITLE
Logsupport

### PR DESCRIPTION
**Pull request by nirmalvp**
1. Removed all logging from the main code using decorators
   1. The interface selection is now circular. Pressing up from the first item takes cursor to the last item and vice versa

### CODE REVIEW

**Comment by neo1691** [comment]
Hi, this is great. 

I was trying to do the same. I have already implemented a small version of argparse, but I was not able to pass the values of --log-file and set the log level accordingly.

Also I tried to make sure that unless a -d or --debug option is not passed, the log should only be in the log file and not on the console, which I failed at doing. 

I will now test your code and merge it, and later I will work argparse

**Comment by nirmalvp** [comment]
I think what you are trying to do is already described in the logging - howTo documentation of python

**Comment by nirmalvp** [comment]
And btw I noticed that the code in your developed branch doesn't log anything other than the app. Python file. I have fixed it in this commits

**Comment by neo1691** [comment]
Thanks a lot, Have you checked out the argparse branch? Also I have read the whole documentation twice now. But I am not able to get it right!

**Comment by nirmalvp** [comment]
Can you summarize what you are trying to do and maybe I can help out if I know what is the right approach

**Comment by nirmalvp** [comment]
You want to get the log levels and the log file name through the user cmd args and set logging accordingly? Why not parse the args (which I am sure you know how to) and then use getattr()  builtin function to set these values to logger object?

**Comment by neo1691** [comment]
I am doing everything, but I think I am not configuring the parent logger properly. And I am not getting time to work on either my code or your PR. I will be doing that in the coming two days.

**Comment by nirmalvp** [comment]
Don't worry about it. I fixed it. The argparse is working just like you described

**Comment by neo1691** [comment]
Ah Great. Thanks I will give it a look

**Comment by nirmalvp** [comment]
That will be an other PR though ☺

**Comment by neo1691** [comment]
Ha Ha. I have time today, I will hopefully merge the code in the codebase.

**Comment by nirmalvp** [comment]
Do you want me to send you a pull request for the argparse branch also?

**Comment by neo1691** [comment]
@nirmalvp is the argparse branch depended on this logsupport branch? I have a few questions regarding logsupport, which I am yet to ask.

**Comment by nirmalvp** [comment]
I modified the code that exist in the argparse branch in your repo. The changes in logsupport branch arent present in my argparse.

**Comment by nirmalvp** [comment]
What questions do you have?

**Comment by neo1691** [comment]
I am just trying to make a decision whether writing a new class, an introducing a new module on the package level is really required against writing two more lines of code in every function. 

I didn't knew about functools and it was very interesting to learn about it, all I want to understand is that should I go with customlog wherein two lines of logger.info are replaced with one line of the decorater @log. 

I wonder which is more pythonic. The official documentation of logging doesn't mention making use of functools, but this is very interesting implementation. Sorry for the delay in replying, working at office, and at the same time preparing for a weekend at Goa!! :P

**Comment by nirmalvp** [comment]
Since logging is a "cross cutting concern", I think the current decorator pattern is a clean implementation than having the logs in every function. Also, helps to maintain the "single responsibility" principle in functions

**Comment by nirmalvp** [comment]
I don't understand the pythonic part of your comment though. Decorators are heavily used in python for scenarios like this. All what functool. Wrap does is maintain the meta data of the original function.

**Comment by nirmalvp** [comment]
How far is Goa from where you live?

**Comment by neo1691** [comment]
@nirmalvp goa is 8 hours from pune. I am leaving now. I will try to merge the code on monday, since I am going to Goa an I'lI will be coming on Monday, so no time to look at your PR now. Sorry, since you have clearly put in a lot of effort for this. 

I can see how decorators can make our life useful. customlog.py is taking care of entry points and exit points from the code, I am wondering if that is really necessary. I am also thinking do we really need to log the entry and exit points of the functions. I am thinking of removing them altogether. 

Let me know what you think about me. 

Life has been a little difficult for me since the last few weeks, I am staying alone all of a sudden and I need to go to the office which is one hour away, so two hours for travelling and then when I come home, I have started to prepare my own food. That is cooking everything from roti, bhaji dal chawal, I haven't been able to attend my laptop since the last two weeks. Trying to sort my life out :) 

Sorry for the long post.

**Comment by neo1691** [comment]
It's almost three months now. Let's get this back on track!
